pwndbg/gdblib/bpoint.py

```diff
@@ -26,3 +26,56 @@ def should_stop(self):
         its return value determines whether the inferior will be stopped.
         """
         return True
+
+
+REGISTERED_BP_EVENTS = set()
+
+
+class BreakpointEvent(gdb.Breakpoint):
+    """
+    Breakpoint class, similar to gdb.Breakpoint, but executes a given callback
+    when, or very shortly after, a the breakpoint is hit, but does not stop
+    the execution of the inferior.
+
+    This allows us to execute code that changes the state of the inferior safely
+    after a breakpoint is hit.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        REGISTERED_BP_EVENTS.add(id(self))
+
+    def delete(self):
+        REGISTERED_BP_EVENTS.remove(id(self))
+        super().delete()
+
+    def on_breakpoint_hit(self):
+        """
+        This function is called whenever this breakpoint is hit in the code.
+        """
+        pass
+
+
+# Attatch ourselves to the event runtime so that we can fire the
+# on_breakpoint_hit() function for all of the breakpoint events that stopped on
+# a given piece of code.
+def _handle_stop(event):
+    if type(event) is not gdb.BreakpointEvent:
+        # We have nothing to do here.
+        return
+
+    print("Handling BPStop")
+    should_continue = True
+    for bp in event.breakpoints:
+        if id(bp) not in REGISTERED_BP_EVENTS:
+            # This breakpoint does not belong to us. We also can't automatically
+            # resume execution after we finish processing our events, because
+            # someone else expects the code to stop here.
+            should_continue = False
+            continue
+        bp.on_breakpoint_hit()
+    if should_continue:
+        gdb.execute("continue")
+
+
+gdb.events.stop.connect(_handle_stop)
```

Comment:
Ideally we want to setup this event only when user requests got tracking. Also this currently makes it so that `Handling BPStop` is printed out and fails some tests

Reply:
Oh, right, I was thinking those could end up being more of a thing we'd use in other places throughout pwndbg. Taking possibly complex actions when a breakpoint is hit, but without really stop the program is fairly generally useful (the UAF tracker draft also does something similar). So I just have it enabled from the get go.

Though this part is a bit problematic, still, considering the drawbacks I outlined in the open questions section. I still think the mechanism itself needs some discussion, and that's why I didn't actually remove any of the diagnostic printing of `Handling BPStop`. I could just remove it, if you think it's best, I just didn't think was much point in doing it right away when the way it's implemented could just change entirely in the next commit.